File: solutions/python/pig-latin/4/pig_latin.py
"""
This module provides a function to translate English sentences into Pig Latin.

The main function, `translate`, takes an input string and transforms each word
into its Pig Latin equivalent by following specific rules for consonants and
vowels. It's useful for fun linguistic exercises or as a playful conversion tool.
"""

from re import sub


def translate(text: str) -> str:
    """Convert a sentence into Pig Latin.

    This function takes a text input and translates each word into Pig Latin.
    The first consonant or consonant cluster is moved to the end of the word,
    and 'ay' is appended. If a word begins with a vowel sound, 'ay' is simply
    added to the end.

    Args:
        text (str): The input string containing words to be translated.

    Returns:
        str: A new string where each word has been transformed into Pig Latin.
    """
    return " ".join(
        sub(
            r"([xy](?=[aeiouy])|s*qu|[bcdfghjklmnpqrstvwz]*)(.*)",
            r"\2\1",
            word,
        )
        + "ay"
        for word in text.split()
    )

File: solutions/javascript/pig-latin/1/pig-latin.js
/**
 * Regular expression to match words starting with a vowel sound.
 * @type {RegExp}
 */
const VOWEL_SOUND_REGEXP = /^([aeiou]|xr|yt)/;

/**
 * Regular expression to match words starting with a consonant sound.
 * @type {RegExp}
 */
const CONSONANT_SOUND_REGEXP = /^([^aeiou]+(?=y)|[^aeiou]?qu|[^aeiou]+)([a-z]+)/;

/**
 * Translates a single word to Pig Latin.
 * If the word starts with a vowel sound, 'ay' is appended.
 * If the word starts with a consonant sound, the consonant(s) are moved to the end and 'ay' is appended.
 *
 * @param {string} word - The word to translate.
 * @returns {string} - The translated word in Pig Latin.
 */
function translateWord(word) {
  if (VOWEL_SOUND_REGEXP.test(word)) {
    return `${word}ay`;
  }

  const newWord = word.replace(CONSONANT_SOUND_REGEXP, '$2$1');
  return `${newWord}ay`;
}

/**
 * Translates a sentence to Pig Latin.
 *
 * @param {string} english - The sentence to translate.
 * @returns {string} - The translated sentence in Pig Latin.
 */
export const translate = (english) => {
  return english.split(' ').map(translateWord).join(' ');
};
